core/src/main/scala/kafka/server/BrokerMetadataEventManager.scala

```diff
@@ -0,0 +1,198 @@
+//        case scramRecord: ScramRecord => handleScramRecord(scramRecord)
+//        case featureLevelRecord: FeatureLevelRecord => handleFeatureLevelRecord(featureLevelRecord)
+        case _ => handleUnexpected(metadataMessage)
+      }
+    } catch {
+      case e: FatalExitError => throw e
+      case e: Exception => handleError(metadataMessage, e)
+    }
+  }
+
+  private def handleBrokerRecord(brokerRecord: BrokerRecord): Unit = {
+    replicaManager.maybeUpdateMetadataCache(brokerRecord)
+  }
+
+//  private def handleFenceBrokerRecord(fenceBrokerRecord: FenceBrokerRecord): Unit = {
+//    throw new UnsupportedOperationException(s"Unimplemented: $fenceBrokerRecord") // TODO: implement-me
+//  }
+
+  private def handleTopicRecord(topicRecord: TopicRecord): Unit = {
+    throw new UnsupportedOperationException(s"Unimplemented: $topicRecord") // TODO: implement-me
+  }
+
+//  private def handleRemoveTopicRecord(removeTopicRecord: RemoveTopicRecord): Unit = {
+//    throw new UnsupportedOperationException(s"Unimplemented: $removeTopicRecord") // TODO: implement-me
+//  }
+
+  private def handlePartitionRecord(partitionRecord: PartitionRecord): Unit = {
+    throw new UnsupportedOperationException(s"Unimplemented: $partitionRecord") // TODO: implement-me
+  }
+
+  private def handleConfigRecord(configRecord: ConfigRecord): Unit = {
+    throw new UnsupportedOperationException(s"Unimplemented: $configRecord") // TODO: implement-me
+  }
+
+  private def handleIsrChangeRecord(isrChangeRecord: IsrChangeRecord): Unit = {
+    throw new UnsupportedOperationException(s"Unimplemented: $isrChangeRecord") // TODO: implement-me
+  }
+
+  private def handleAccessControlRecord(accessControlRecord: AccessControlRecord): Unit = {
+    throw new UnsupportedOperationException(s"Unimplemented: $accessControlRecord") // TODO: implement-me
+  }
+
+//  private def handleDelegationTokenRecord(delegationTokenRecord: DelegationTokenRecord): Unit = {
+//    throw new UnsupportedOperationException(s"Unimplemented: delegationTokenRecord") // TODO: implement-me
+//  }
+
+//  private def handleScramRecord(scramRecord: ScramRecord): Unit = {
+//    throw new UnsupportedOperationException(s"Unimplemented: scramRecord") // TODO: implement-me
+//  }
+
+//  private def handleFeatureLevelRecord(featureLevelRecord: FeatureLevelRecord): Unit = {
+//    throw new UnsupportedOperationException(s"Unimplemented: featureLevelRecord") // TODO: implement-me
+//  }
+
+  def handleUnexpected(metadataMessage: RaftMessage): Unit = {
+    error(s"Ignoring unexpected metadata message=$metadataMessage")
+  }
+
+  def handleError(metadataMessage: RaftMessage, e: Exception): Unit = {
+    error(s"Error when handling metadata message=$metadataMessage", e)
+  }
+
+  class BrokerMetadataEventThread(name: String) extends ShutdownableThread(name = name, isInterruptible = false) {
+    logIdent = s"[BrokerMetadataEventThread] "
+
+    override def doWork(): Unit = {
```

Comment:
In the interest of efficiency, we should be pulling batches of records off of the queue, rather than single records.  That should be fairly easy to do, though.

Reply:
It would be be easy to use [drainTo(Collection, int)](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/BlockingQueue.html#drainTo-java.util.Collection-int-).  But I believe we spoke yesterday about a related issue, which is that we may not want to expose intermediate state in some cases -- i.e. that we may end up defining a transactional batch of records.  If I simply introduce `drainTo()` then I don't think it ads much except maybe a performance gain -- I would still be processing the individual messages individually and exposing each of the new states.  Are you suggesting I introduce a transactional batch?  I would be excited to add that, but I don't want to go ahead with it without confirmation.

Reply:
It wasn't too difficult to put something rough together for metadata transactions (i.e. don't expose intermediate state), so I pushed a commit with something we can discuss.  Take a look at `BrokerMetadataBasis` and let me know what you think.

Reply:
We now drain to a non-blocking buffer queue.